### PR TITLE
Spring Cleanup

### LINKED ISSUE #26: Error in VSTS - The "Message" parameter is not supported by the "Error" task.
When running in VSTS I get the following errors on PCL Profile 111 projects in a solution mixed with `netstandard1.1` and `netstandard2.0`.

MS Build parameters: `/t:pack /p:PackageVersion=$(Build.BuildNumber) /p:NuGetBuildTasksPackTargets="workaround"`

Error:

`d:\a\1\s\packages\MSBuild.Sdk.Extras.1.1.0\build\NuGet.Workarounds.1.x.targets(12,12): error MSB4064: The "Message" parameter is not supported by the "Error" task. Verify the parameter exists on the task, and it is a settable public instance property. `

### LINKED ISSUE #32: WPF: Adding xmlns with clr-namespace to XAML breaks build with cryptic error message
Dear Oren
Today I tried converting a bare WPF Project, i.E. one that was created by VS 15.5.7 New -> Project -> WPF App (.Net Framework) to the new csproj sdk format. I got stuck with:

> 1>C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(419,45): error MSB4057: The target "CoreCompile" does not exist in the project.

Removing the following line from MainWindow.xaml fixes the build:

> xmlns:local="clr-namespace:SdkCsprojWpfSpike"

ATM I'm quite perplexed how this can have any effect on the build outcome, but I'd guess it's about the WPF two-pass build...?!

--------------

Notably, (not) having the same line @ App.xaml doesn't matter.

---------------

Sample Project
---------

Using your sample project I can reproduce the same issue.
The steps I performed:

## 1 ./Samples/WpfApplication/MainWindow.xaml
insert the following as Line 8:

> xmlns:local="clr-namespace:WpfApplication"

image

## 2 ./Samples/WpfApplication/WpfApplication.csproj
I've had to change the target framework to **net471** - because I don't have net462 installed.

## 3 Open Csproj & Build
.. and you'll see
> 1>C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319\Microsoft.WinFx.targets(419,45): error MSB4057: The target "CoreCompile" does not exist in the project.

------------------

I'd be happy providing a PR for the readme/wiki page to document this, at the least. Any directions for a header, where to place it the document?

### LINKED ISSUE #35: 'The target "Build" does not exist in the project' for PCL profiles 2 and 328
Target framework strings:

* `portable-net4+sl40+win8+wp70` (Profile2)
* `portable-net4+sl50+win8+wpa81+wp8` (Profile328)

Command line and output:

```
PS> dotnet pack /bl
Microsoft (R) Build Engine version 15.5.180.51428 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files\dotnet\sdk\2.1.4\MSBuild.dll /bl /Logger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,C:\Program Files\dotnet\sdk\2.1.4\dotnet.dll /m /restore /t:pack /v:m .\Structurizr.Annotations.csproj
  Restoring packages for C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\Structurizr.Annotations.csproj...
  Generating MSBuild file C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\obj\Structurizr.Annotations.csproj.nuget.g.props.
  Generating MSBuild file C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\obj\Structurizr.Annotations.csproj.nuget.g.targets.
  Restore completed in 542.19 ms for C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\Structurizr.Annotations.csproj.
C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\Structurizr.Annotations.csproj : error MSB4057: The target "Build" does not exist in the project.
C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\Structurizr.Annotations.csproj : error MSB4057: The target "Build" does not exist in the project.
  Structurizr.Annotations -> C:\Users\ccharabaruk\Repos\structurizr-dotnet\Structurizr.Annotations\bin\Debug\netstandard1.0\Structurizr.Annotations.dll
PS>
```

Binlog output: [msbuild.binlog.zip](https://github.com/onovotny/MSBuildSdkExtras/files/1764070/msbuild.binlog.zip)

### LINKED ISSUE #53: Update samples to the VS 2017 15.6 way of things.
Any chance of updating the to the new SDK way of doing things, I noticed the class library has already been updated be useful for the application csproj to be updated also.

### LINKED ISSUE #64: dotnet sdk import issues
Hi @onovotny. First I'd like to thank you for putting this together, this has helped me tremendously so far.

I am in the middle of converting a solution of 4 projects to use the new csproj format using the .NET SDK. One of the requirements of this project is also to convert to using dotnet for all build commands. I have to make sure I get this to build in two different environments: using VS 2017 locally, for debugging, and using the dotnet sdk on the build server. This one WPF project is giving me significant pains. It currently builds in VS 2017 just fine, but when I am running the command `dotnet build --verbosity normal /p:Configuration=Release` I am getting the message:  `C:\Program Files\dotnet\sdk\2.1.105\Sdks\Microsoft.NET.Sdk\Sdk\Sdk.targets(41,3): error MSB4019: The imported project "C:\Program Files\dotnet\sdk\2.1.105\15.0\Bin\Microsoft.CSharp.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk. [*.csproj]`

This is an example of my .csproj with some unnecessary stuff removed for the sake of anonymity (marked with a @@).

```XML
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <TargetFramework>net45</TargetFramework>
    <!--<RuntimeIdentifier>win7-x64</RuntimeIdentifier>-->
    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
    <DocumentationFile>bin\$(Configuration)\@@.xml</DocumentationFile>
    <OutputPath>bin\$(Configuration)</OutputPath>
    <StartupObject>@@.App</StartupObject>
    <ApplicationIcon>Resources\@@.ico</ApplicationIcon>
    <OutputType>WinExe</OutputType>
    <LanguageTargets>$(MSBuildToolsPath)\Microsoft.CSharp.targets</LanguageTargets>
    <!--<CSharpTargets>$(MSBuildToolsPath)\Microsoft.CSharp.targets</CSharpTargets>-->
    <!--<CSharpTargets Condition="'$(CSharpTargets)' == ''">$(MSBuildToolsPath)\Microsoft.CSharp.targets</CSharpTargets>-->
  </PropertyGroup>
  <!-- Workaround for lack of XAML support using new project format. -->
  <Import Project="$(MSBuildSDKExtrasTargets)" Condition="Exists('$(MSBuildSDKExtrasTargets)')" />
  <!--<Import Project="$(CSharpTargets)" />-->
  <!--<Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />-->
  <!--<Import Project="$(BaseIntermediateOuputPath)$(OriginalProjectName).*.props" Condition=" '$(MSBuildProjectName)' != '$(OriginalProjectName)' and '$(ImportProjectExtensionProps)' != 'false' and exists('$(BaseIntermediateOuputPath)')" />-->
  <ItemGroup>
    <PackageReference Include="FontAwesome.WPF" Version="4.5.0.8" />
    <PackageReference Include="MSBuild.Sdk.Extras" Version="1.2.2" PrivateAssets="all" />
    <ProjectReference Include="..\@@.csproj" />
    <ProjectReference Include="..\@@.csproj" />
    <ProjectReference Include="..\@@.csproj" />
    <Reference Include="System" />
    <Reference Include="System.Configuration" />
    <Reference Include="System.Data" />
    <Reference Include="System.IO.Compression" />
    <Reference Include="System.Net.Http" />
    <Reference Include="System.Xml" />
    <Reference Include="System.Core" />
    <Reference Include="System.Data.DataSetExtensions" />
    <Reference Include="System.Xaml" />
    <Reference Include="WindowsBase" />
    <Reference Include="PresentationCore" />
    <Reference Include="PresentationFramework" />
  </ItemGroup>
  <!--<Import Project="$(BaseIntermediateOuputPath)$(OriginalProjectName).*.targets" Condition=" '$(MSBuildProjectName)' != '$(OriginalProjectName)' and '$(ImportProjectExtensionProps)' == 'true' and exists('$(BaseIntermediateOuputPath)')" />-->
  <ItemGroup>
    <Resource Include="Resources\*.png;Resources\*.ico" />
    <EmbeddedResource Include="Resources\*.dll" />
  </ItemGroup>
</Project>
```

Now I can't confirm this to be a result of your package, as it might be a hidden error otherwise. I can comment out the MSBuildSDKExtrasTargets Import but I get stuck with xaml compilation errors which your package seems to help. I also no longer get the error when commented though so that leads me to believe something is creating a bad import from somewhere.

I'm slowly pulling my hair out with this. Do you have any ideas?

### LINKED ISSUE #66: Build failing when including WPF items
Today I had a really weird issue when converting a test project (of Catel) to the new project structure. It was building fine, but once I did a few (seemed simple) changes, the build suddenly broke:

```
PublicApiFacts.cs(11,11): error CS0246: The type or namespace name 'ApiApprover' could not be found (are you missing a using directive or an assembly reference?) [C:\CI_WS\Ws\110208\Source\Catel\src\Catel.Tests\Catel.Tests_fh3rqzzc_wpftmp.csproj]
```

Visual Studio also started complaining that it was unable to see the item, but it was there! Then I decided to remove a test user control I added (to check for xaml resource existence). And suddenly the error disappeared.

Not sure what's happening under the hood, but I think it's weird that adding a WPF user control breaks the build on a totally different item (e.g. a NuGet package with file references that are automatically linked in):

image

### LINKED ISSUE #67: Using MSBuild Sdk extras for UWP apps
I am currently converting everything to this new magic stuff, and have a question about UWP apps (not libraries).

Would it make sense to add something like this in by default or do you think it should be done per project (but I'd like to keep my csproj as clean as possible)?

```
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\x86\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <NoWarn>1591;1998;2008</NoWarn>
    <DebugType>full</DebugType>
    <PlatformTarget>x86</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x86'">
    <OutputPath>bin\x86\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>1591;1998;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>x86</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>true</UseDotNetNativeToolchain>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|ARM'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\ARM\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <NoWarn>;2008</NoWarn>
    <DebugType>full</DebugType>
    <PlatformTarget>ARM</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|ARM'">
    <OutputPath>bin\ARM\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>ARM</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>true</UseDotNetNativeToolchain>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">
    <DebugSymbols>true</DebugSymbols>
    <OutputPath>bin\x64\Debug\</OutputPath>
    <DefineConstants>DEBUG;TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <NoWarn>;2008</NoWarn>
    <DebugType>full</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">
    <OutputPath>bin\x64\Release\</OutputPath>
    <DefineConstants>TRACE;NETFX_CORE;WINDOWS_UWP</DefineConstants>
    <Optimize>true</Optimize>
    <NoWarn>;2008</NoWarn>
    <DebugType>pdbonly</DebugType>
    <PlatformTarget>x64</PlatformTarget>
    <UseVSHostingProcess>false</UseVSHostingProcess>
    <ErrorReport>prompt</ErrorReport>
    <Prefer32Bit>true</Prefer32Bit>
    <UseDotNetNativeToolchain>true</UseDotNetNativeToolchain>
  </PropertyGroup>
```

### LINKED ISSUE #83: Readme.MD file is referencing old version
https://github.com/onovotny/MSBuildSdkExtras/blob/master/README.md

It's referencing a old version of the project in the 15.6+ section, it's mentioning 1.2.2

Should be 1.5.4

### PR DESCRIPTION
**Pull request by Nirmal4G**
### These are the issues that are fixed in this PR

**Done**

- [x] Closes #26
- [x] Closes #32
- [x] Closes #35
- [x] Closes #53
- [x] Closes #64
- [x] Closes #66
- [x] Closes #67
- [x] Closes #83

Will be updated along the way.

### CODE REVIEW

**Comment by Nirmal4G** [comment]
There are still patches pending, so until all the patches have been applied, I'll hold off on any renames (file or otherwise)

TestProjects are also pending. I already logged the changes that you want to make. I'll make them when I finalize this PR.

**Comment by Nirmal4G** [comment]
Also Why have you removed `Directory.Build.rsp`? It is useful for local builds!

File: README.md [hunk 1]
```diff
@@ -1,136 +1,134 @@
...
+#### Using the SDK
 
-MyGet CI feed: `https://myget.org/F/msbuildsdkextras/api/v3/index.json`
+1. Create a new project
+    - .NET Core console app or .NET Standard class library.
+    - With your exisiting SDK-style project.
+    - With the templates in the repo's [TestProjects](/TestProjects) folder.
 
-[![MSBuild.Sdk.Extras](https://img.shields.io/myget/msbuildsdkextras/v/MSBuild.Sdk.Extras.svg)](https://myget.org/gallery/msbuildsdkextras)
+2. Replace `Microsoft.NET.Sdk` with `MSBuild.Sdk.Extras` to the project's top-level `Sdk` attribute.
```
Comment (onovotny) [inline README.md]: This only works if there's a global.json with the SDK version in it. otherwise you need the /1.3.4 syntax.

File: README.md [hunk 2]
```diff
@@ -1,136 +1,139 @@
...
 
-## Using the package (VS 2017 15.6+)
+3. You have to tell MSBuild that the `Sdk` should resolve from NuGet by
+    - Adding a `global.json` containing the Sdk name and version.
+    - Appending a version info to the `Sdk` attribute value.
```
Comment (Nirmal4G) [inline README.md]: Is this ok?

File: Source/MSBuild.Sdk.Extras/Build/FrameworkResolvers/MSBuild.NETPortable.v4.5.targets [hunk 3]
```diff
@@ -1,51 +1,51 @@
...
-  </PropertyGroup>
-
+﻿<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
-  <PropertyGroup Condition="$(_SdkShortFrameworkProfileCount) == 2 and
-      $(_SdkShortFrameworkProfile.Contains('+net45+')) and
-      ($(_SdkShortFrameworkProfile.Contains('+win+')) or $(_SdkShortFrameworkProfile.Contains('+win8+')) or $(_SdkShortFrameworkProfile.Contains('+win80+')))">
+  <PropertyGroup Condition="$(_ShortFrameworkProfileCount) == 2 AND
+      $(_ShortFrameworkProfile.Contains('+net45+')) AND
+      ($(_ShortFrameworkProfile.Contains('+win+')) OR $(_ShortFrameworkProfile.Contains('+win8+')) OR $(_ShortFrameworkProfile.Contains('+win80+')))">
```
Comment (onovotny) [inline Source/MSBuild.Sdk.Extras/Build/FrameworkResolvers/MSBuild.NETPortable.v4.5.targets]: Please do not rename all of the internal variables like this. I explicitly want them prefixed as _Sdk* so I don't accidently step-on others.

Comment (Nirmal4G) [reply Source/MSBuild.Sdk.Extras/Build/FrameworkResolvers/MSBuild.NETPortable.v4.5.targets]: They don't have any equivalent in the Main SDK. Also I have tested these properties, for a long time, none of the internal properties clash with the Main SDK. Some properties do blend with the Main SDK without affecting the build.

Comment (Nirmal4G) [reply Source/MSBuild.Sdk.Extras/Build/FrameworkResolvers/MSBuild.NETPortable.v4.5.targets]: It was easy for me to debug the targets since the naming conventions are similar with Main SDK.

**Comment by onovotny** [comment]
Is this PR ready for review or is there more?

**Comment by Nirmal4G** [comment]
There are a little bit of hotfix patches to apply!
Other than that it's ready.

The renames will be done at the last commit just before merging the PR, other than that you can review the functionality and the structure for now. I'll improve them based on your feedback.

**Comment by onovotny** [comment]
What is the purpose of changing all of the `TargetFrameworkVersion` to remove the `v` only to set it again later? That seems like it'd be confusing.

File: Source/MSBuild.Sdk.Extras/Build/MSBuild.NET.Inference.targets [hunk 4]
```diff
@@ -114,4 +106,10 @@
     <Import Condition="'$(TargetFrameworkVersion)' != ''" Project="$(MSBuildThisFileDirectory)FrameworkResolvers\MSBuild.NETPortable.v$(TargetFrameworkVersion).targets"/>
```
Comment (Nirmal4G) [inline Source/MSBuild.Sdk.Extras/Build/MSBuild.NET.Inference.targets]: It's so that this import should work. I could have done it a different way but this seems better in some sense, my reason is that any msbuild files within `FrameworkResolvers` folder can only have version number only, since any logic added afterwards involving setting `TFV` shouldn't worry about having `v` prefix, It just works at the end.

**Comment by Nirmal4G** [comment]
https://github.com/onovotny/MSBuildSdkExtras/pull/80#discussion_r193490911

File: Source/MSBuild.Sdk.Extras/Build/MSBuild.NET.Inference.targets (hunk 4 above)
Comment (onovotny) [reply Source/MSBuild.Sdk.Extras/Build/MSBuild.NET.Inference.targets]: Can it be    `<Import Condition="'$(TargetFrameworkVersion)' != ''" Project="$(MSBuildThisFileDirectory)FrameworkResolvers\MSBuild.NETPortable.$(TargetFrameworkVersion).targets"/> ` and leave the v in the variable? Is there a case where it'd only be the version?

Reassigning the variable like this may lead to subtle errors and make following the values in the trace harder.

File: Source/MSBuild.Sdk.Extras/Build/MSBuild.NET.Inference.targets [hunk 5]
```diff
@@ -114,4 +106,10 @@
...
   </ImportGroup>
 
+  <PropertyGroup>
+    <_TargetFrameworkVersionWithoutV>$(TargetFrameworkVersion)</_TargetFrameworkVersionWithoutV>
+    <_TargetFrameworkVersionWithV>v$(TargetFrameworkVersion)</_TargetFrameworkVersionWithV>
+    <TargetFrameworkVersion>$(_TargetFrameworkVersionWithV)</TargetFrameworkVersion>
+  </PropertyGroup>
```
Comment (Nirmal4G) [inline Source/MSBuild.Sdk.Extras/Build/MSBuild.NET.Inference.targets]: > Can it be `
<Import Condition="'$(TargetFrameworkVersion)' != ''" Project="$(MSBuildThisFileDirectory)FrameworkResolvers\MSBuild.NETPortable.$(TargetFrameworkVersion).targets"/>`

We can do that.

> Reassigning the variable like this may lead to subtle errors and make following the values in the trace harder.

But here this is the single point of change where the prefixing of `v` actually happens, everywhere else it's just version, so tracing is easy. Also when the import reaches platform targets, we can be confident that any assigned `TFV` will be valid `vX.X` format!

File: Source/MSBuild.Sdk.Extras/DefaultItems/MSBuild.NET.ImplicitPackages.props [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      Packing should include the implicit package dependency for non-executables.
+    -->
+    <PackageReference Update="Microsoft.NETCore.UniversalWindowsPlatform" PrivateAssets="All" Publish="false"
+      Condition="'$(OutputType)' != 'Exe' OR '$(OutputType)' != 'AppContainerExe'"/>
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(DisableImplicitFrameworkReferences)' != 'true' AND '$(TargetFrameworkIdentifier)' == 'Tizen'">
+    <PackageReference Include="Tizen.NET" Version="$(_ImplicitPlatformPackageVersion)" IsImplicitlyDefined="true"/>
+    <PackageReference Update="Tizen.NET" PrivateAssets="All" Publish="false" Condition="'$(OutputType)' != 'Exe'"/>
+  </ItemGroup>
```
Comment (Nirmal4G) [inline Source/MSBuild.Sdk.Extras/DefaultItems/MSBuild.NET.ImplicitPackages.props]: Are the UWP and Tizen packages private for all output types or only for the library-like output types?

File: TestProjects/Windows-Desktop-C#/Console-Application/Console-Application.csproj [hunk 7]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="MSBuild.Sdk.Extras">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net45</TargetFramework>
+    <ProjectGuid>{F15AFA64-1CA5-486A-B9EA-B05377DF2FEC}</ProjectGuid>
```
Comment (Nirmal4G) [inline TestProjects/Windows-Desktop-C#/Console-Application/Console-Application.csproj]: Do we need the `ProjectGuid` tags?

I added them since `sln` changes the `Guid` when a project is removed and re-added. If there's any other reason like Visual Studio needs it, we can keep it, or we can remove them!

**Comment by Nirmal4G** [comment]
I'm Finalizing this PR. I'll create a separate PR for renames, tests, etc...

Please look at the above comments.

**Comment by Nirmal4G** [comment]
Done. You can merge this anytime. Please do look at the above comments.

I'll create a separate PR for feedback changes, renames, other test projects etc...
